chain/chain/src/test_utils.rs

```diff
@@ -1127,3 +1127,65 @@ impl ChainGenesis {
         }
     }
 }
+
+#[cfg(test)]
+mod test {
+    use super::KeyValueRuntime;
+    use crate::types::ReceiptList;
+    use crate::RuntimeAdapter;
+    use borsh::BorshSerialize;
+    use near_primitives::hash::{hash, CryptoHash};
+    use near_primitives::receipt::Receipt;
+    use near_primitives::types::NumShards;
+    use near_store::test_utils::create_test_store;
+    use rand::Rng;
+
+    impl KeyValueRuntime {
+        fn naive_build_receipt_hashes(&self, receipts: &[Receipt]) -> Vec<CryptoHash> {
+            let mut receipts_hashes = vec![];
+            for shard_id in 0..self.num_shards() {
+                let shard_receipts: Vec<Receipt> = receipts
+                    .iter()
+                    .filter(|&receipt| {
+                        self.account_id_to_shard_id(&receipt.receiver_id) == shard_id
+                    })
+                    .cloned()
+                    .collect();
+                receipts_hashes
+                    .push(hash(&ReceiptList(shard_id, &shard_receipts).try_to_vec().unwrap()));
+            }
+            receipts_hashes
+        }
+    }
+
+    fn test_build_receipt_hashes_with_num_shard(num_shards: NumShards) {
+        let store = create_test_store();
+        let runtime_adapter = KeyValueRuntime::new_with_validators(
+            store,
+            vec![(0..num_shards).map(|i| format!("test{}", i)).collect()],
+            1,
+            num_shards,
+            10,
+        );
+        let create_receipt_from_receiver_id =
+            |receiver_id| Receipt::new_balance_refund(&receiver_id, 0);
+        let mut rng = rand::thread_rng();
+        let receipts = (0..100)
+            .map(|_| {
+                let random_number = rng.gen_range(0, 1000);
+                create_receipt_from_receiver_id(format!("test{}", random_number))
+            })
+            .collect::<Vec<_>>();
+        assert_eq!(
```

Comment:
This tests the new version of the code produces the same result, which is good. But perhaps we should also tests that it is faster too? That is the purpose of this change after all and it would be interesting to quantify the speedup.

Reply:
I added some basic performance check to the test in https://github.com/nearprotocol/nearcore/pull/3235/commits/4481e964c205cb0f943bb6012edd15ed61dcd28d

The new version is quite a lot faster!